### PR TITLE
feat(): first round stackblitz integration

### PR DESCRIPTION
**Pull request by ghost**
## Description

Demo via Stackblitz api integration

### What's included?

- Demo via Stackblitz api integration

#### Test Steps

- [ ] `npm run serve`
- [ ] Verify there are stackblitz demos for all except: Loading mask, Code Editor, Text Editor. (Will be resolved in future)

### CODE REVIEW

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 1]
```diff
@@ -1,22 +1,206 @@
...
+  polyfills: string;
+}
 
 @Component({
   selector: 'demo-component',
   styleUrls: ['./demo.component.scss'],
   templateUrl: './demo.component.html',
 })
 export class DemoComponent {
+  private readonly uniqueChaCha: string = 'Cha-234c8b6d-591e-4140-9c82-8c36a0709afb-Cha';
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Let's add more descriptive names 😆

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 2]
```diff
@@ -1,22 +1,206 @@
...
+  private readonly uniqueYahYah: string = 'Yah-f3cc5d58-5956-446a-abee-a8fb70193768-Yah';
   @Input() demoId: string;
   @Input() demoTitle: string;
+  @Input() hideStackBlitzDemo: boolean = false;
   viewCode: boolean = false;
   typescriptFile: string;
   htmlFile: string;
   stylesFile: string;
+  dependencies: object = new Object();
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: let's keep it  simple and use `{}`

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 3]
```diff
@@ -1,22 +1,206 @@
...
+  indexFile: string = `
+  <!doctype html>
+  <html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <title>ChaCha</title>
+    <base href="/">
+  </head>
+  <body>
+    <Cha-234c8b6d-591e-4140-9c82-8c36a0709afb-Cha></Cha-234c8b6d-591e-4140-9c82-8c36a0709afb-Cha>
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Let's reference uniqueChaCha uniqueYahYah in all the places these keys are duplciated.

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 4]
```diff
@@ -1,22 +1,206 @@
 import { Component, Input } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
-import { forkJoin, Observable, Subscriber } from 'rxjs';
+import sdk from '@stackblitz/sdk';
 import { catchError, map } from 'rxjs/operators';
+import { forkJoin, Observable, Subscriber, of } from 'rxjs';
+import * as data from '../../../../../package.json';
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Let's give this a more descriptive name, like `packageJson`. Also we can de-structure if dependencies is all we use.

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 5]
```diff
@@ -1,22 +1,206 @@
...
+    providers: [],
+    bootstrap: [Yah-f3cc5d58-5956-446a-abee-a8fb70193768-Yah]
+  })
+  export class AppModule { }
+    `;
+
+  constructor(private _http: HttpClient) {
+    this.dependencies = data.dependencies;
+    // has additional themes for respective demos to work
+    this.dependencies['@covalent/core'] = '3.0.0-rc.2';
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: We should place the version in one var and ref that. Also let's just use an object destructing instead of assigning each dependency 1 by 1

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 6]
```diff
@@ -1,22 +1,206 @@
...
+      flavoredMarkdownThemeScss: this._http.get(`platform/flavored-markdown/_flavored-markdown-theme.scss`, {
+        responseType: 'text',
+      }),
+      teradataThemeScss: this._http.get(`assets/demos/_teradata-branding.scss`, { responseType: 'text' }),
+      polyfills: this._http.get(`assets/demos/polyfills.ts`, { responseType: 'text' }),
+    })
+      .pipe(
+        map((responses: IDemoFiles) => {
+          return {
+            typescript: responses.typescript,
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Can restructure all these keys from the response

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 7]
```diff
@@ -1,22 +1,206 @@
...
+        this.appModuleFile = this.appModuleFile.replace(this.uniqueChaCha, this.demoId);
+        this.appModuleFile = this.appModuleFile.replace(new RegExp(this.uniqueYahYah, 'g'), demoComponentName);
+
+        // TODO: remove this after markdown has been update and pushed out with the respective import
+        demo.markdownThemeScss = demo.markdownThemeScss.replace(
+          "@import '../../../node_modules/@angular/material/theming';",
+          "@import '~@angular/material/theming';",
+        );
+
+        const project: any = {
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: We can type this

File: src/app/content/components/component-demos/breadcrumbs/demos/breadcrumbs-demo.module.ts [hunk 8]
```diff
@@ -7,7 +7,7 @@ import { MatDividerModule } from '@angular/material/divider';
 
 import { BreadcrumbsDemoBasicComponent } from './breadcrumbs-demo-basic/breadcrumbs-demo-basic.component';
 import { BreadcrumbsDemoCustomSeparatorComponent } from './breadcrumbs-demo-custom-separator/breadcrumbs-demo-custom-separator.component';
-import { BreadcrumbsResizeBasicComponent } from './breadcrumbs-demo-resize/breadcrumbs-demo-resize.component';
+import { BreadcrumbsDemoResizeComponent } from './breadcrumbs-demo-resize/breadcrumbs-demo-resize.component';
```
Comment (christianmemije) [inline src/app/content/components/component-demos/breadcrumbs/demos/breadcrumbs-demo.module.ts]: Can you clarify why this renaming was necessary?

File: src/app/content/components/component-demos/data-table/demos/data-table-demo-basic/data-table-demo-basic.component.ts [hunk 9]
```diff
@@ -20,10 +20,120 @@ export class DataTableDemoBasicComponent implements OnInit {
...
+      .afterClosed()
+      .subscribe((value: any) => {
+        if (value !== undefined) {
+          row[name] = value;
+        }
+      });
+  }
+
+  async ngOnInit(): Promise<void> {
+    this.data = [
```
Comment (christianmemije) [inline src/app/content/components/component-demos/data-table/demos/data-table-demo-basic/data-table-demo-basic.component.ts]: This data already got moved in #1659 so need to merge in latest

File: src/app/content/components/component-demos/breadcrumbs/demos/breadcrumbs-demo.module.ts (hunk 8 above)
Comment (stevenov7) [reply src/app/content/components/component-demos/breadcrumbs/demos/breadcrumbs-demo.module.ts]: The component name must match the filename/selector `breadcrumbs-demo-resize` since that's the key i use to create the component name on the fly and inject it into the demo's app.module.

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 10]
```diff
@@ -1,22 +1,198 @@
...
   styleUrls: ['./demo.component.scss'],
   templateUrl: './demo.component.html',
 })
-export class DemoComponent {
-  @Input() demoId: string;
+export class DemoComponent implements AfterViewInit {
+  private readonly uniqueFileNamePlaceholder: string = 'uniqueFileNamePlaceholder-234c8b6d-591e-4140-9c82-8c36a0709afb';
+  private readonly uniqueComponentNamePlaceholder: string =
+    'uniqueComponentNamePlaceholder-f3cc5d58-5956-446a-abee-a8fb70193768';
+  @ViewChild('content') private content: any;
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Can we type this?

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 11]
```diff
@@ -1,22 +1,198 @@
-import { Component, Input } from '@angular/core';
+import { Component, Input, ViewChild, AfterViewInit } from '@angular/core';
 import { HttpClient } from '@angular/common/http';
-import { forkJoin, Observable, Subscriber } from 'rxjs';
-import { catchError, map } from 'rxjs/operators';
+import sdk from '@stackblitz/sdk';
+import { forkJoin, Observable, Subscriber, of } from 'rxjs';
+import * as packageJson from '../../../../../package.json';
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Can you do `import { version } from '../../../../../package.json';` ?

Comment (stevenov7) [reply src/app/components/shared/demo-tools/demo.component.ts]: Yah i thinks so as of some version of typescript a while back. https://mariusschulz.com/blog/importing-json-modules-in-typescript

Comment (stevenov7) [reply src/app/components/shared/demo-tools/demo.component.ts]: Tried it, doesn't work

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 12]
```diff
@@ -1,22 +1,198 @@
...
+  indexFile: string = `
+  <!doctype html>
+  <html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <title>ChaCha</title>
+    <base href="/">
+  </head>
+  <body>
+    <uniqueFileNamePlaceholder-234c8b6d-591e-4140-9c82-8c36a0709afb></uniqueFileNamePlaceholder-234c8b6d-591e-4140-9c82-8c36a0709afb>
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Can we reference uniqueFileNamePlaceholder and uniqueComponentNamePlaceholder in these template strings?

File: src/app/components/shared/demo-tools/demo.component.ts [hunk 13]
```diff
@@ -1,22 +1,198 @@
...
+    ],
+    providers: [],
+    bootstrap: [uniqueComponentNamePlaceholder-f3cc5d58-5956-446a-abee-a8fb70193768]
+  })
+  export class AppModule { }
+    `;
 
-  constructor(private _http: HttpClient) {}
+  constructor(private _http: HttpClient, private _cdr: ChangeDetectorRef) {
+    this.dependencies = packageJson.dependencies;
```
Comment (christianmemije) [inline src/app/components/shared/demo-tools/demo.component.ts]: Let's cleanup this by using object destructing. 
```ts
this.dependencies = {
...packageJson.dependencies, 
@covalent/core: this.version,
....
}
```

File: src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts [hunk 14]
```diff
@@ -12,14 +12,5 @@ export class FlavoredMarkdownDemoBasicComponent {
     - [x] My checkbox
     - [x] My second checkbox
     - [ ] My empty checkbox
-
```
Comment (christianmemije) [inline src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts]: Why this change?

File: src/app/content/components/component-demos/shared.module.ts [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    CovalentMapEchartsModule,
+    CovalentPieEchartsModule,
+    CovalentSankeyEchartsModule,
+    CovalentScatterEchartsModule,
+    CovalentTreeEchartsModule,
+    CovalentTreemapEchartsModule,
+    CovalentWordcloudEchartsModule,
+  ],
+  providers: [],
+  entryComponents: [
```
Comment (christianmemije) [inline src/app/content/components/component-demos/shared.module.ts]: Entry components are no longer necessary as of ng 9.

File: src/app/content/components/component-demos/theme.scss [hunk 16]
```diff
@@ -0,0 +1,244 @@
+@import '~@angular/material/theming';
```
Comment (christianmemije) [inline src/app/content/components/component-demos/theme.scss]: Can you leave a comment at the top of this file stating the difference between this and the original theme.scss, so we  know what needs to be updated in the future. Also maybe add a comment to theme.scss stating that this file has to be kept in sync with this one. Or actually why not just pull in the original theme.scss and replace the parts of the file that needs to be changed, versus having 2 copies.

File: src/app/components/shared/demo-tools/demo.component.ts (hunk 13 above)
Comment (stevenov7) [reply src/app/components/shared/demo-tools/demo.component.ts]: for some reason doing it this way with the spread operator breaks the stackblitz. so going to keep it as is for now.

    this.dependencies = {
      ...packageJson.dependencies, 
      // has additional themes for respective demos to work
      '@covalent/core': this.version,
      '@covalent/dynamic-forms': this.version,
  
      '@covalent/markdown': this.version,
      '@covalent/flavored-markdown': this.version,
      '@covalent/markdown-navigator': this.version,
      '@covalent/code-editor': this.version,
      '@covalent/http': this.version,
      '@covalent/highlight': this.version,
      // TODO: doesn't appear stackblitz api support nightly dependencies - needs follow up for loading mask demo
      // this.dependencies['@covalent/experimental'] = "git+https://github.com/Teradata/covalent-experimental-nightly.git";
      // "@covalent/experimental": "git+https://github.com/Teradata/covalent-experimental-nightly.git",
      '@covalent/echarts': this.version,
      '@types/echarts': '*',
      'zrender': '4.3.0', // needed for echarts
  
      '@td-vantage/ui-platform': '*'
      }

File: src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts (hunk 14 above)
Comment (stevenov7) [reply src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts]: b/c of this error and the thing is was asking for was not being exposed to access it.
Screen Shot 2020-03-23 at 2 02 33 PM

File: src/app/content/components/component-demos/shared.module.ts (hunk 15 above)
Comment (stevenov7) [reply src/app/content/components/component-demos/shared.module.ts]: Not sure why but via Stackblitz w/ ng 9 it was asking for them if we don't add them here.

File: src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts (hunk 14 above)
Comment (christianmemije) [reply src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts]: Well the thing is that we need this as part of the demo, so we should solve that issue instead.

File: src/app/content/components/component-demos/shared.module.ts (hunk 15 above)
Comment (christianmemije) [reply src/app/content/components/component-demos/shared.module.ts]: That's odd. Is there a configuration we are missing somewhere? If not, can you add a comment explaining the reason for adding those.

Comment (stevenov7) [reply src/app/content/components/component-demos/shared.module.ts]: Hmm after looking into it more. Seems like it's an issue with stackblitz angular-template (they need to update it on their end w/ latest @angular/cli and angular-devkit/build-angular) and ivy support. Can you help take a look in case i'm missing something @christianmemije

File: src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts (hunk 14 above)
Comment (stevenov7) [reply src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts]: *same comment https://github.com/Teradata/covalent/pull/1682#discussion_r397993619

Comment (christianmemije) [reply src/app/content/components/component-demos/flavored-markdown/demos/flavored-markdown-demo-basic/flavored-markdown-demo-basic.component.ts]: Need to probably add a tsconfig, and properly reference it in the angular.json 

https://stackblitz.com/github/ocombe/cli-ivy-test
